### PR TITLE
Port code to System.Text.Json

### PR DESCRIPTION
**Pull request by oschwald**

### CODE REVIEW

File: MaxMind.MinFraud.UnitTest/TestData/factors-response.json [hunk 1]
```diff
@@ -56,12 +56,18 @@
     "location": {
       "accuracy_radius": 96,
       "latitude": 51.5142,
-      "local_time": "2012-04-13T00:20:50+01:00",
+      "local_time": "2012-04-13T00:20:50-01:00",
       "longitude": -0.0931,
       "time_zone": "Europe\/London"
     },
+    "maxmind": {},
```
Comment (oschwald) [inline MaxMind.MinFraud.UnitTest/TestData/factors-response.json:63]: This is just to make testing easier. It's presence seems harmless.

File: MaxMind.MinFraud.UnitTest/TestData/factors-response.json [hunk 2]
```diff
@@ -91,8 +113,10 @@
       "domain": "in-addr.arpa",
       "ip_address": "152.216.7.110",
       "is_anonymous": true,
+      "is_anonymous_proxy": true,
       "is_anonymous_vpn": true,
       "is_hosting_provider": true,
+      "is_legitimate_proxy": false,
```
Comment (oschwald) [inline MaxMind.MinFraud.UnitTest/TestData/factors-response.json:119]: Similar to the above.

File: MaxMind.MinFraud.UnitTest/WebServiceClientTest.cs [hunk 3]
```diff
@@ -157,8 +152,8 @@ public async Task Test200WithNoBody()
 
             var exception = await Record.ExceptionAsync(async () => await client.InsightsAsync(request));
             Assert.NotNull(exception);
-            Assert.IsType<HttpException>(exception);
-            Assert.Contains("Received a 200 response for minFraud Insights but there was no message body",
+            Assert.IsType<MinFraudException>(exception);
```
Comment (oschwald) [inline MaxMind.MinFraud.UnitTest/WebServiceClientTest.cs:155]: This exception changed as we are using the combined extension method `ReadFromJsonAsync` for reading the content and deserializing it. I don't think the change is matters.

File: MaxMind.MinFraud.UnitTest/MaxMind.MinFraud.UnitTest.csproj [hunk 4]
```diff
@@ -29,21 +28,16 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="Microsoft.Bcl.HashCode" Version="1.1.0" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.8.0" />
-    <PackageReference Include="MockHttpSigned" Version="1.3.0-netstandard-alpha3" />
+    <PackageReference Include="RichardSzalay.MockHttp" Version="6.0.0" />
```
Comment (oschwald) [inline MaxMind.MinFraud.UnitTest/MaxMind.MinFraud.UnitTest.csproj:33]: We switched from an ancient internal fork back to upstream.

File: MaxMind.MinFraud.UnitTest/MaxMind.MinFraud.UnitTest.csproj [hunk 5]
```diff
@@ -3,11 +3,10 @@
   <PropertyGroup>
     <Description>Test project for minFraud API</Description>
     <VersionPrefix>0.6.0</VersionPrefix>
-    <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT' ">net452;net46;net5.0;netcoreapp2.1;netcoreapp3.1</TargetFrameworks>
+    <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT' ">net461;net5.0;netcoreapp2.1;netcoreapp3.1</TargetFrameworks>
     <TargetFrameworks Condition=" '$(OS)' != 'Windows_NT' ">net5.0;netcoreapp2.1;netcoreapp3.1</TargetFrameworks>
     <AssemblyName>MaxMind.MinFraud.UnitTest</AssemblyName>
     <AssemblyOriginatorKeyFile>../MaxMind.snk</AssemblyOriginatorKeyFile>
-    <SignAssembly>true</SignAssembly>
```
Comment (oschwald) [inline MaxMind.MinFraud.UnitTest/MaxMind.MinFraud.UnitTest.csproj:10]: We are no longer using the friend assembly feature so we don't need to sign the unit tests.

File: MaxMind.MinFraud/Exception/HttpException.cs [hunk 6]
```diff
@@ -13,6 +14,7 @@ namespace MaxMind.MinFraud.Exception
     ///     by the web service itself. As such, it is a IOException instead of a
     ///     MinFraudException.
     /// </summary>
+    [Serializable]
```
Comment (oschwald) [inline MaxMind.MinFraud/Exception/HttpException.cs:17]: The changes to the exceptions are unrelated to the JSON changes. I just made the same changes that I did in the GeoIP2 library, following the recommended pattern.

File: MaxMind.MinFraud/Request/CustomInputs.cs [hunk 7]
```diff
@@ -23,6 +23,12 @@ private CustomInputs()
         {
         }
 
+        /// <summary>
+        ///     The custom inputs to be sent to the web service.
+        /// </summary>
+        [JsonExtensionData]
+        public IDictionary<string, object> Inputs => new ReadOnlyDictionary<string, object>(_inputs);
```
Comment (oschwald) [inline MaxMind.MinFraud/Request/CustomInputs.cs:30]: Unfortunately, `System.Text.Json` doesn't support making this an `IReadOnlyDictionary`.

File: MaxMind.MinFraud/WebServiceClient.cs [hunk 8]
```diff
@@ -257,44 +249,47 @@ private static async Task Handle4xxStatus(HttpResponseMessage response)
...
 
-            if (string.IsNullOrEmpty(content))
+            if (content == null || content.Length == 0)
             {
                 throw new HttpException(
                     $"Received a {status} error for {uri} with no body",
                     response.StatusCode, uri);
             }
 
+            var decoded = Encoding.UTF8.GetString(content);
```
Comment (horgh) [inline MaxMind.MinFraud/WebServiceClient.cs:261]: Not that it'll ever be an issue in practice, but do we know it's UTF-8?

**Review by horgh — CHANGES_REQUESTED** [review]
👍

File: MaxMind.MinFraud/Exception/HttpException.cs [hunk 9]
```diff
@@ -65,6 +67,31 @@ public HttpException(string message, HttpStatusCode httpStatus, Uri? uri, System
...
         }
+        /// <summary>
+        ///     Constructor for deserialization.
+        /// </summary>
+        /// <param name="info"></param>
+        /// <param name="context"></param>
+        protected HttpException(SerializationInfo info, StreamingContext context) : base(info, context)
+        {
+            HttpStatus = (HttpStatusCode)(info.GetValue("MaxMind.MinFraud.Exception.HttpException.HttpStatus", typeof(HttpStatusCode))
+                ?? throw new SerializationException("Unexcepted null HttpStatus value"));
```
Comment (horgh) [inline MaxMind.MinFraud/Exception/HttpException.cs:78]: "Unexpected". Looks like there are other instances too.

File: MaxMind.MinFraud/Request/Event.cs [hunk 10]
```diff
@@ -71,35 +71,36 @@ public sealed class Event
...
         /// </summary>
-        [JsonProperty("time")]
+        [JsonPropertyName("time")]
         public DateTimeOffset? Time { get; }
 
         /// <summary>
         /// The type of event being scored.
         /// </summary>
-        [JsonProperty("type")]
+        [JsonConverter(typeof(EnumMemberValueConverter<EventType>))]
```
Comment (horgh) [inline MaxMind.MinFraud/Request/Event.cs:102]: Do one of our tests cover serializing this using this?

**Review by horgh — CHANGES_REQUESTED** [review]

File: MaxMind.MinFraud/Request/Event.cs (hunk 10 above)
Comment (oschwald) [reply MaxMind.MinFraud/Request/Event.cs:102]: Yeah, we do test the serialization of the request sent to the mock web service. However, a direct test sounds worthwhile. I'll add that.

**Review by horgh — APPROVED** [review]

**Pull request merged**